File: internal/domain/template.go
package domain

import "time"

// TemplateTier is where a template lives. Unlike a skill's tier this is not a
// precedence — two templates never apply at once, because exactly one is chosen
// at kickoff and then the template is done. It is only ownership.
type TemplateTier string

const (
	// TemplateGlobal ships with the binary and is refreshed at boot. Nobody
	// owns it; editing one forks a team copy.
	TemplateGlobal TemplateTier = "global"
	// TemplateTeam belongs to a team and is theirs alone.
	TemplateTeam TemplateTier = "team"
)

// TemplateSource splits the global tier in two, and the split is what stops an
// upgrade eating somebody's work. Both kinds are global — visible to every team
// on the instance, owned by no team — but only one of them is refreshed from the
// binary at boot.
type TemplateSource string

const (
	// TemplateSourceBuiltin ships in the binary. The boot-time loader owns this
	// row: it rewrites the body on upgrade, and nothing else may edit it.
	TemplateSourceBuiltin TemplateSource = "builtin"
	// TemplateSourceUser was written on this instance — a team template, or a
	// global one added by the operator through the API. The loader never
	// touches it.
	TemplateSourceUser TemplateSource = "user"
)

// Limits the service enforces. The two matching fields are capped because they
// are what `template_list` returns to every kickoff — the body is not, so it is
// allowed to be a real document.
const (
	// TemplateCriterionMax bounds `when_to_use` and `when_not_to_use`. Six
	// templates times two lines is the whole of what an agent reads before
	// choosing, so it has to stay small.
	TemplateCriterionMax = 240
	// TemplateBodyMax is generous on purpose: a kickoff methodology is read
	// once, in full, by a model that has just started. It is not the always-in-
	// context budget a skill description lives under.
	TemplateBodyMax = 24000
)

// TemplateSkill is one entry of a template's skill list, resolved for display.
// `Missing` is the interesting one: a built-in naming an unknown skill fails the
// boot, but nothing stops a team skill being deleted after a template named it.
type TemplateSkill struct {
	Slug        string `json:"slug"`
	Name        string `json:"name,omitempty"`
	Description string `json:"description,omitempty"`
	Ambient     bool   `json:"ambient,omitempty"`
	Missing     bool   `json:"missing,omitempty"`
}

type Template struct {
	ID string `json:"id"`
	// TeamID is empty for a global template. Tier carries the same fact in the
	// shape callers branch on.
	TeamID string       `json:"team_id,omitempty"`
	UserID string       `json:"user_id,omitempty"`
	Slug   string       `json:"slug"`
	Name   string       `json:"name"`
	Tier   TemplateTier `json:"tier"`
	// Source says whether this row came out of the binary or was written here.
	// A client needs it because the two are different products to a reader: one
	// is refreshed on upgrade and can only be forked, the other was written by
	// somebody on this server and can be edited by whoever may edit it.
	Source TemplateSource `json:"source"`
	// Description is one line for a picker. WhenToUse and WhenNotToUse are what
	// an agent actually matches on.
	Description  string `json:"description"`
	WhenToUse    string `json:"when_to_use"`
	WhenNotToUse string `json:"when_not_to_use"`
	// Body is omitted from every list response: it is the document, and
	// shipping it in a list would put every methodology on the instance into a
	// kickoff that needs one.
	Body string `json:"body,omitempty"`
	// Skills are slugs attached to the research this template starts.
	Skills []string `json:"skills"`
	// ForkedFrom is the global slug this copies, when it is a fork.
	ForkedFrom string `json:"forked_from,omitempty"`
	// SkillsResolved is set on a single-template read, so a reader sees what
	// the methodology will attach rather than a list of slugs. Derived.
	SkillsResolved []TemplateSkill `json:"skills_resolved,omitempty"`
	// ResearchCount is how many researches this has started, among those the
	// caller can read. Derived, and computed only on a single read — omitted
	// rather than sent as zero from a list, because a list row reading
	// "used 0 times" for a template in daily use is worse than saying nothing.
	ResearchCount int `json:"research_count,omitempty"`
	// BodyWords is a size hint for a reader deciding whether to open it, and it
	// is an estimate: a list cannot afford to fetch the body, so it is counted
	// in SQL by whitespace rather than by tokenising. Derived, never stored.
	BodyWords int       `json:"body_words"`
	Version   int       `json:"version"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

File: frontend/__mocks__/template.ts
/** Templates as the API returns them, for stories and for anything that needs a
 *  catalogue without a server behind it.
 *
 *  `tier` and `source` are two different facts and the fixtures keep both: the
 *  tier says who owns it, the source says whether an upgrade rewrites it. The
 *  global tier holds both kinds — what ships in the binary, and what the
 *  operator of an instance added through the API.
 *
 *  The other thing these fixtures encode is that a list row and a single read
 *  are not the same object. `GET /api/templates` never sends the body, the
 *  research count or the resolved skills; only `GET /api/templates/{id}` does.
 *  So the list fixtures below carry none of the three, and a story that shows a
 *  row "started 9 researches" is showing a payload the server does not send. */

/** One entry of `skills_resolved`. `missing` is the one worth rendering: a team
 *  skill can be deleted after a template named it. */
export interface MockTemplateSkill {
  slug: string
  name?: string
  description?: string
  ambient?: boolean
  missing?: boolean
}

export interface MockTemplate {
  id: string
  slug: string
  name: string
  tier: 'global' | 'team'
  source: 'builtin' | 'user'
  /* Never omitted by the API. They arrive as '' when unset, which is why the
     components branch on falsiness rather than on the key being absent. */
  description: string
  when_to_use: string
  when_not_to_use: string
  skills: string[]
  /** Counted in SQL from the body, so a list can show a size without the text. */
  body_words: number
  version: number
  created_at: string
  updated_at: string
  /** Team tier only — a global template belongs to nobody. */
  team_id?: string
  /** Who wrote it. Absent on what ships in the binary. */
  user_id?: string
  /** The global slug this copies. A fork keeps its parent's slug, so the two
   *  are equal on a real fork. */
  forked_from?: string

  /* Single read only (`GET /api/templates/{id}`). Never in a list. */
  body?: string
  research_count?: number
  skills_resolved?: MockTemplateSkill[]
}

const at = (day: number) => `2025-08-${String(day).padStart(2, '0')}T09:14:00Z`

/** What ships. Two of the ten — enough to see a list, not a catalogue dump.
 *  No `user_id`: the loader wrote these, not a person. */
export const mockGlobalTemplates: MockTemplate[] = [
  {
    id: 'g1', slug: 'technology-comparison', name: 'Technology comparison',
    tier: 'global', source: 'builtin', version: 3, body_words: 812,
    created_at: at(1), updated_at: at(9),
    description: 'Choose between named alternatives and be able to defend the choice afterwards.',
    when_to_use:
      'Use when a decision is being made between named tools, libraries or platforms and the reasoning will have to survive review.',
    when_not_to_use:
      'Not when the choice is already made and what is wanted is justification, and not for a market scan — that is a landscape.',
    skills: ['evidence-grading'],
  },
  {
    id: 'g2', slug: 'user-interview-study', name: 'User interview study',
    tier: 'global', source: 'builtin', version: 2, body_words: 954,
    created_at: at(1), updated_at: at(4),
    description: 'Findings traceable to what real people actually said.',
    when_to_use:
      'Use when the question is what people do and why, and the evidence has to come from talking to them rather than from analytics.',
    when_not_to_use:
      'Not for measuring how many — that is a survey. Not when nobody can be recruited in the time available.',
    skills: ['structured-interviewing', 'evidence-grading'],
  },
]

/** Added on this server by whoever runs it: global, but not ours. Written
 *  through the API by an operator, so it has a `user_id` and version 1. */
export const mockHouseTemplates: MockTemplate[] = [
  {
    id: 'h1', slug: 'house-diligence', name: 'House diligence',
    tier: 'global', source: 'user', user_id: 'usr_ops', version: 1, body_words: 430,
    created_at: at(6), updated_at: at(6),
    description: 'The way this company runs a diligence, before anything else is decided.',
    when_to_use:
      'Use when a deal, a supplier or a partnership is being assessed here, and the answer has to go to the committee.',
    when_not_to_use: 'Not for a technical evaluation — use the comparison methodology for that.',
    skills: ['evidence-grading'],
  },
]

/** A team's own, and its edited copy of a shipped one. A fork keeps the parent's
 *  slug, which is why `slug` and `forked_from` match on the second. */
export const mockTeamTemplates: MockTemplate[] = [
  {
    id: 't1', slug: 'our-playbook', name: 'Our discovery playbook',
    tier: 'team', source: 'user', team_id: 'team-1', user_id: 'usr_ada',
    version: 5, body_words: 601, created_at: at(2), updated_at: at(14),
    description: 'How this team opens a discovery, including the two questions we always forget.',
    when_to_use:
      'Use when starting discovery on a new customer problem in this team, before any solution has been sketched.',
    when_not_to_use: 'Not for a delivery post-mortem.',
    skills: ['structured-interviewing'],
  },
  {
    id: 't2', slug: 'literature-review', name: 'Literature review (ours)',
    tier: 'team', source: 'user', team_id: 'team-1', user_id: 'usr_ada',
    forked_from: 'literature-review', version: 2, body_words: 770,
    created_at: at(3), updated_at: at(11),
    description: 'The shipped review methodology with our exclusion rules bolted on.',
    when_to_use:
      'Use when the answer is expected to exist already in published work and the job is to find, exclude and summarise it.',
    when_not_to_use: 'Not when the field has no published work worth the search.',
    skills: ['evidence-grading'],
  },
]

/** No matching line: an agent will never choose it, and the row has to say so.
 *  A write cannot produce this any more — `when_to_use` is required — but a row
 *  that predates the check, or a built-in file whose frontmatter omits the key,
 *  still arrives looking like this. */
export const mockTemplateWithoutCriteria: MockTemplate = {
  id: 'x1', slug: 'silent', name: 'Untriggered methodology',
  tier: 'team', source: 'user', team_id: 'team-1', user_id: 'usr_ada',
  version: 1, body_words: 88, created_at: at(7), updated_at: at(7),
  description: '', when_to_use: '', when_not_to_use: '', skills: [],
}

/** Both criteria within a few runes of the 240 the service allows, with an
 *  unbroken product name in one. This is the width the row and the detail page
 *  are actually designed for. */
export const mockTemplateLongCriteria: MockTemplate = {
  id: 'x2', slug: 'platform-migration-readiness',
  name: 'Platform migration readiness assessment for long-lived internal services',
  tier: 'team', source: 'user', team_id: 'team-2', user_id: 'usr_kim',
  version: 1, body_words: 1840, created_at: at(5), updated_at: at(15),
  description: 'Whether a service can be moved at all, before anyone argues about where to.',
  when_to_use:
    'Use when a long-lived internal service is being considered for a move onto another platform — kubernetes-on-bare-metal-with-cilium-and-longhorn, a managed runtime, or somebody else entirely — and the decision has to survive the team',
  when_not_to_use:
    'Not when the move is already committed and dated, and not when the service is small enough that two people could rewrite it in a fortnight — the assessment then costs more than the migration it is meant to de-risk in the end',
  skills: ['evidence-grading', 'structured-interviewing'],
}

/** One template as a single read: body, research count and resolved skills, none
 *  of which a list carries. */
export const mockTemplateDetail: MockTemplate = {
  ...mockGlobalTemplates[0]!,
  research_count: 4,
  skills_resolved: [
    {
      slug: 'evidence-grading', name: 'Grading evidence',
      description: 'Use when recording a claim that rests on a source, or when two sources disagree.',
    },
  ],
  body: [
    '# Technology comparison',
    '',
    'This research ends in a choice somebody has to live with.',
    '',
    '## Before you propose anything',
    '',
    '1. **What breaks if this decision is wrong, and when would you find out?**',
    '2. **What are you not allowed to change?** Licence terms, an existing contract,',
    '   the language the team already writes.',
    '',
    '## Working rules',
    '',
    '**Criteria are fixed and weighted before any candidate is named.** Otherwise',
    'they get retrofitted around whichever option somebody already likes.',
    '',
    '**"Keep what we have" is candidate zero**, scored like the others.',
  ].join('\n'),
}

/** A single read whose skill list has rotted: the team skill it names was
 *  deleted, so the server reports it missing rather than dropping it. */
export const mockTemplateDetailMissingSkill: MockTemplate = {
  ...mockTemplateDetail,
  id: 'g1-missing',
  skills: ['evidence-grading', 'house-rules'],
  skills_resolved: [
    ...(mockTemplateDetail.skills_resolved ?? []),
    { slug: 'house-rules', missing: true },
  ],
}
